src/linear/ridge_regression.rs

```diff
@@ -0,0 +1,323 @@
+
+        let (n, p) = x.shape();
+
+        if n <= p {
+            return Err(Failed::fit(&format!(
+                "Number of rows in X should be >= number of columns in X"
+            )));
+        }
+
+        let y_column = M::from_row_vector(y.clone()).transpose();
+
+        let (w, b) = if parameters.normalize {
+            let (scaled_x, col_mean, col_std) = Self::rescale_x(x)?;
+            let x_t = scaled_x.transpose();
+            let x_t_y = x_t.matmul(&y_column);
+            let mut x_t_x = x_t.matmul(&scaled_x);
+
+            for i in 0..p {
+                x_t_x.add_element_mut(i, i, parameters.alpha);
+            }
+
+            let mut w = match parameters.solver {
+                RidgeRegressionSolverName::Cholesky => x_t_x.cholesky_solve_mut(x_t_y)?,
+                RidgeRegressionSolverName::SVD => x_t_x.svd_solve_mut(x_t_y)?,
+            };
+
+            for i in 0..p {
+                w.set(i, 0, w.get(i, 0) / col_std[i]);
+            }
+
+            let mut b = T::zero();
+
+            for i in 0..p {
+                b += w.get(i, 0) * col_mean[i];
+            }
+
+            let b = y.mean() - b;
+
+            (w, b)
+        } else {
+            let x_t = x.transpose();
+            let x_t_y = x_t.matmul(&y_column);
+            let mut x_t_x = x_t.matmul(x);
+
+            for i in 0..p {
+                x_t_x.add_element_mut(i, i, parameters.alpha);
+            }
+
+            let w = match parameters.solver {
+                RidgeRegressionSolverName::Cholesky => x_t_x.cholesky_solve_mut(x_t_y)?,
+                RidgeRegressionSolverName::SVD => x_t_x.svd_solve_mut(x_t_y)?,
+            };
+
+            (w, T::zero())
+        };
+
+        Ok(RidgeRegression {
+            intercept: b,
+            coefficients: w,
+            solver: parameters.solver,
+        })
+    }
+
+    fn rescale_x(x: &M) -> Result<(M, Vec<T>, Vec<T>), Failed> {
+        let col_mean = x.mean(0);
+        let col_std = x.std(0);
+
+        for i in 0..col_std.len() {
+            if (col_std[i] - T::zero()).abs() < T::epsilon() {
+                return Err(Failed::fit(&format!(
+                    "Cannot rescale constant column {}",
+                    i
+                )));
+            }
+        }
+
+        let mut scaled_x = x.clone();
+        scaled_x.scale_mut(&col_mean, &col_std, 0);
+        Ok((scaled_x, col_mean, col_std))
+    }
+
+    /// Predict target values from `x`
+    /// * `x` - _KxM_ data where _K_ is number of observations and _M_ is number of features.
+    pub fn predict(&self, x: &M) -> Result<M::RowVector, Failed> {
+        let (nrows, _) = x.shape();
+        let mut y_hat = x.matmul(&self.coefficients);
+        y_hat.add_mut(&M::fill(nrows, 1, self.intercept));
+        Ok(y_hat.transpose().to_row_vector())
+    }
+
+    /// Get estimates regression coefficients
+    pub fn coefficients(&self) -> M {
+        self.coefficients.clone()
```

Comment:
Do we need to clone here?

Reply:
with the current interface we do need to clone. If I decide to change this interface I should also change interfaces of `LinearRegression` and `LogisticRegression` as well for consistency reasons. 

I agree that returning a reference here is an overkill